_posts/2016-03-04-hunting-down-a-slow-rails-request.markdown

```diff
@@ -0,0 +1,97 @@
+---
+layout: post
+title: "Hunting Down a Slow Rails Request"
+date: 2016-03-04 11:03:00
+categories: ["rails", "performance"]
+author: "mauro-oto"
+---
+
+We have recently began using [Skylight](https://www.skylight.io) in production
+for one of our client's Rails applications, as an attempt to try to improve the
+performance of some of the more critical API endpoints.
+
+Skylight reports on both time taken, with a breakdown of time taken per SQL
+query, and also on object allocations per request. I noticed an unusually large
+amount of allocated objects for one request:
+
+![Skylight report](/blog/assets/images/high-object-allocation.png)
+
+This request would take anywhere from 400ms to 3000ms to respond, which is WAY,
+way too long.
+
+<!--more-->
+
+To debug this, I first captured the `String` object allocation count locally to
+see how it behaved:
+
+```ruby
+Started GET "/email_accounts.json?results_per_page=30" for 127.0.0.1 at 2016-03-02 13:04:22 +0000
+Processing by EmailAccountsController#index as JSON
+  Parameters: {"results_per_page"=>"30"}
+
+[[String, 150299], [Arel::Nodes::SqlLiteral, 5], [Arel::Nodes::BindParam, 4], [ActiveSupport::StringInquirer, 1]]
+### JSON rendering happens here ###
+[[String, 707982], [ActiveSupport::JSON::Encoding::JSONGemEncoder::EscapedString, 2922], [Arel::Nodes::SqlLiteral, 6], [ActiveSupport::StringInquirer, 1]]
+
+Completed 200 OK in 5860ms (Views: 11.8ms | ActiveRecord: 564.9ms)
+```
+
+To get the amount of allocated strings, I used the following snippet:
+
+```ruby
+strings = {}
+ObjectSpace.each_object(String) { |str| strings[str.class] += 1 }
+pp strings
+```
+
+As you can see, there are 700k+ allocated strings, and this increase in object
```

Comment:
@mauro-oto I'd separate this into "As you can see, there are 700k+ allocated strings. It's very likely that this amount of allocations is the culprit for the slow performance."